Mylin/1617 support spatial profile line polyline

acdo2002 opened the pull request:
Fixed #1617 
The corresponding backend [PR1020](https://github.com/CARTAvis/carta-backend/pull/1020)

acdo2002 committed:
match pam's protobuf

acdo2002 committed:
set width to SpatialProfileWidgetStore.ts

acdo2002 committed:
remove the block line region in dropdown Region

acdo2002 committed:
feed xMax and x value to plotData

acdo2002 committed:
remove case RegionsType.CLOSED & case RegionsType.CLOSED_AND_POINT in RegionSelectorComponent.tsx

acdo2002 committed:
add LineRegionSampleWidth in SpatialProfileWidgetStore.ts

acdo2002 committed:
plot line region & without x wcs coordinate

acdo2002 committed:
add computation tab

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

acdo2002 committed:
remove import CARTA

acdo2002 committed:
add width

acdo2002 committed:
define bp3-input scss

acdo2002 committed:
attach the line region plotdata

acdo2002 committed:
Match Pam's protobuf

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

Commit:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

acdo2002 committed:
refactory SpatialProfileWidgetStore.ts

acdo2002 committed:
add Region.Type POLYLINE

acdo2002 committed:
reformat

acdo2002 committed:
resolve merge conflict in RegionSelectorComponent.tsx

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

acdo2002 committed:
remove xMin and modify x

Commit:
fix the xMin and xMax for line/polyline region

acdo2002 committed:
fix polyline show top Axis

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

Commit:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

kswang1029 commented:
I have had a preview of this branch and found few issues that need to be addressed:
1. when the region is point or cursor, the x label of the profile plot is "X coordinate" or "Y coordinate" (with an implicit unit of "pixel". When the region is line (not polyline), the x label should be revised to offset with a unit (eg Offset (arcsec). The actual unit (arcsec, arcmin, or degree) is determined by the angular size of the line region and the backend should already have a handle on this based on the implementation of pv generator. When the region is polyline, the x label should be "Distance" with a unit. I suspect we need to update protobuf message SPATIAL_PROFILE_DATA. 
2. If we update the line or polyline region and before the new results come in, we need to have visual indication that the backend is working on it. Similar to region spectral profilers, we should set the old profile dimmer.
3. the red vertical line should disappear when the region type is line and polyline. (as a future enhancement, we should display a marker along the line or polyline region and sync its movement with the red vertical line in the spatial profile plot)
4. if the active line region is deleted, the spatial profiler has an odd behavior displaying the new active region.

https://user-images.githubusercontent.com/20819712/163362347-1ed4616d-c794-492b-b1c2-4d040ac6bd2e.mov

5. the title of the spatial profiler widget should be more dynamic based on what is being displayed. Before the support of line and polyline region, it is like "X Profile: Cursor". Now this should be like "Profile: region 2" when region 2 is a line region or a polyline region. 
6. some UI elements in the smoothing tab of the settings dialog are not rendered properly.
![Screen Shot 2022-04-14 at 18 04 45](https://user-images.githubusercontent.com/20819712/163363384-1c1166d9-597f-4327-997d-b5ab1a848212.png)

Commit:
fixed bp3-input width in SpatialProfilerSettingsPanelComponent.scss

acdo2002 committed:
add opacity condition to cursorX marker

acdo2002 committed:
npm run reformat

acdo2002 committed:
change coordinateString if the region is line/polyline

acdo2002 committed:
modify bp3-input width in SpatialProfilerSettingsPanelComponent.scss

acdo2002 committed:
revise width to 88

acdo2002 commented:
@kswang1029 
I have shown the current frontend to Tien-Hao and already discussed with Tien-Hao.
1. may need the modification of ICD message, I leave it later.
2. may need the modification of ICD message, I leave it later.
3. Fixed, I have let the red vertical line to be transparent if the region is line/polyline. (the enhancement please issue it in the future)
4. It is because once the region (no matter the region is line/polyline) is deleted, the cursor will show the previous cursor spatial profile data. If you move the cursor (slightly around the region) before you delete the region, the cursor will show a relatively correct value. We think this is another issue.
5. Fixed the coordinate string, now it will show "Profile: Region # number" in the spatial profiler widget
6. Fixed the scss.

kswang1029 commented:
> 4. It is because once the region (no matter the region is line/polyline) is deleted, the cursor will show the previous cursor spatial profile data. If you move the cursor (slightly around the region) before you delete the region, the cursor will show a relatively correct value. We think this is another issue.

issue filed at https://github.com/CARTAvis/carta-frontend/issues/1831

TienHao committed:
removed x/y coordinate for line regions

* added POINT_AND_LINES in RegionTypes for Spatial Profiler

* minor refactoring

acdo2002 committed:
set xlabel when line and polyline

acdo2002 committed:
update protobuf

acdo2002 committed:
npm run reformat

acdo2002 committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

TienHao committed:
modified effectiveRegionId of active region

Commit:
set x-label name

Commit:
set the xrange for line/polyline regions

Commit:
reformat

kswang1029 commented:
@acdo2002 the displayed unit in the cursor info text string needs to be sync'd to the actual unit of the line/polyline region. Also the WCS part needs to be removed when it is line or polyline region. 
![Screen Shot 2022-04-26 at 13 57 48](https://user-images.githubusercontent.com/20819712/165231692-6af94ac0-88d0-429d-abfc-66718e59dd1d.png)

kswang1029 commented:
@acdo2002 when it is line or polyline profile being displayed, the header of the exported tsv file needs to be updated and include the line or polyline information. At the moment it is still a point region. 
![Screen Shot 2022-04-26 at 14 10 36](https://user-images.githubusercontent.com/20819712/165233658-8bc3659b-8544-4003-8e97-82d5c3c2c79a.png)

kswang1029 commented:
@acdo2002 spotted a frontend crash. If we explicitly set the image dropdown to an image (not "Active") and explicitly set the region dropdown to a line region (not "Active"), then if we delete the region, frontend will crash.
![Screen Shot 2022-04-26 at 14 38 31](https://user-images.githubusercontent.com/20819712/165237622-0406989c-4b77-49b8-a58c-a8a48fc26cff.png)

Commit:
Fixed Cursor Info Image unit

TienHao committed:
fixed smoothing profile of line and polyline regions

TienHao committed:
fixed line region delete bug, and modified region data of line/polyline region in export data header

kswang1029 commented:
@TienHao as we have a new “compute” tab in the settings dialog, please add a placeholder for the in-app help.

kswang1029 commented:
@TienHao Please help to refine the cursor info:
1. when the region is line or polyline, we don't need to show world coordinate in the cursor info
2. please switch to scientific notation to show distance or offset. %.5e should be good enough. 
3. Replace "Image" to "Offset" (for line region) or "Distance" (for polyline region). 

![Screen Shot 2022-04-27 at 11 41 38](https://user-images.githubusercontent.com/20819712/165435690-dd8a0872-4e47-42c3-a3dc-7c21964ca60d.png)

kswang1029 commented:
@TienHao The tsv header and file name need another tweak. When the profile is from line or polyline region, we should say "Region Profile" instead of "X profile" in the first line of the tsv file. Similar for the file name of the tsv.
![Screen Shot 2022-04-27 at 11 55 05](https://user-images.githubusercontent.com/20819712/165436945-17328e20-d8f4-4cbe-a5bb-d94b90c60b18.png)

kswang1029 commented:
@TienHao please enable auto y-scale when zooming for a line or polyline profile, just like point or cursor region profile.

kswang1029 commented:
@TienHao The title of the settings dialog needs a tweak too. I suggest we unify it to "Spatial Profile Settings"

In addition, the show mean/rms function does not work for line/polyline profile. 
![Screen Shot 2022-04-27 at 13 26 18](https://user-images.githubusercontent.com/20819712/165447207-c39b602d-9c1b-4502-8728-0ef56ef53678.png)

TienHao committed:
code refactoring

TienHao committed:
added help content, fixed bugs

* refined cursor info

* modified exportData header and title of settings dialog

* enable auto y-scale zooming

* fixed Mean/RMS

TienHao committed:
Merge branch 'dev' into mylin/1617-support_spatial_profile_line_polyline

TienHao committed:
removed wcs data in cursor info of line/polyline regions, and fixed local decimated profile

kswang1029 commented:
@YuHsuan-Hwang @veggiesaurus I have had brief tests during the draft PR stage. Now it is ready for code review. I will continue my tests before approving.

kswang1029 reviewed (approved):
@TienHao @acdo2002 this works nicely. 👍

YuHsuan-Hwang commented on src/components/SpatialProfiler/SpatialProfilerComponent.tsx:
```diff
@@ -384,10 +437,11 @@ export class SpatialProfilerComponent extends React.Component<WidgetProps> {
+        const xLabel = this.lineAxis ? `${this.lineAxis.label} (${this.lineAxis.unit ?? ""})` : `${isXProfile ? "X" : "Y"} coordinate`;
         const imageName = appStore.activeFrame ? appStore.activeFrame.filename : undefined;
-        const plotName = `${isXProfile ? "X" : "Y"} profile`;
+        const plotName = `${this.lineAxis ? "" : isXProfile ? "X" : "Y"} profile`;
```
remove extra space for line regions in filename
<img width="677" alt="image" src="https://user-images.githubusercontent.com/43841102/167063125-d74418f4-6796-447a-8d43-ffa6d2707827.png">

acdo2002 replied:
Fixed!

YuHsuan-Hwang commented on src/stores/Frame/RegionStore.ts:
```diff
@@ -239,6 +239,8 @@ export class RegionStore {
             case CARTA.RegionType.POINT:
                 return `Point (pixel) [${center}]`;
+            case CARTA.RegionType.LINE:
+                return `Line (pixel) [[${this.controlPoints[0].x}, ${this.controlPoints[0].y}], [${this.controlPoints[1].x}, ${this.controlPoints[1].y}]]`;
```
Shall we fix the precision to 6 and add unit `pix`? Shall we show "[Invalid]" if the value does not exist?

acdo2002 replied:
I add the lineProperties, making the format similar to polygonProperties & polylineProperties.

YuHsuan-Hwang commented on src/stores/Frame/RegionStore.ts:
```diff
@@ -250,6 +252,13 @@ export class RegionStore {
+            case CARTA.RegionType.POLYLINE:
+                let polylineProperties = "Polyline (pixel) [";
+                this.controlPoints.forEach((point, index) => {
+                    polylineProperties += isFinite(point.x) && isFinite(point.y) ? `[${toFixed(point.x, 6)}, ${toFixed(point.y, 6)}]` : "[Invalid]";
```
Shall we add unit `pix`?

acdo2002 replied:
Fixed!

YuHsuan-Hwang reviewed (changes requested):
The changes look good to me. only very minor issues.

Commit:
fixed minor issue

Commit:
npm run reformat for RegionStore.ts

Commit:
update protobuf

Commit:
Match to Pam's protobuf

acdo2002 commented:
@YuHsuan-Hwang I try my best to solve the merge conflicts (as well as solve the minor issue you suggested), could you check whether this branch is working fine or not?
At least this branch can build successfully on my MacOS 11 laptop, I will try to build on my Ubuntu laptop in the weekend.

YuHsuan-Hwang reviewed (approved):
> @YuHsuan-Hwang I try my best to solve the merge conflicts (as well as solve the minor issue you suggested), could you check whether this branch is working fine or not?
At least this branch can build successfully on my MacOS 11 laptop, I will try to build on my Ubuntu laptop in the weekend.

@acdo2002 looks fine to me 👍

acdo2002 commented:
I have tried to build it on ubuntu 20.04, it looks fine.

Commit:
update Pam's line profile protobuf

veggiesaurus committed:
updated protobuf commit
